### ISSUE TITLE
[MJAVADOC-116] Impossible to aggregate javadoc if snapshot never built

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-javadoc-plugin#451]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** opened **[MJAVADOC-116](https://issues.apache.org/jira/browse/MJAVADOC-116?redirect=false)** and commented

In a multi-module projet, I build an aggregated Javadoc for the site.

The project is built with "mvn clean deploy site-deploy"

When I add a new project, the next build always fails because the javadoc plugin can't find at least one snapshot for the new added module

In the following example, I added a new module tele.persistance:pers-commons, which have never been built before.
Maven tries to download it but it can't find it (never build before).

```
 [INFO] [site:site]
[WARNING] Unable to load parent project from repository: Could not find the model file '/continuum-folders/working-directory/116/../pom.xml'.
[INFO] Skipped "About" report, file "index.html" already exists for the English version.
[ERROR] VM #displayTree: error : too few arguments to macro. Wanted 2 got 0
[ERROR] VM #menuItem: error : too few arguments to macro. Wanted 1 got 0
[INFO] Generate "JavaDocs" report.
[INFO] snapshot tele:commons:1.2.0-alpha-1-SNAPSHOT: checking for updates from mirror.snapshots
[INFO] snapshot tele.persistance:pers-data:1.2.0-alpha-1-SNAPSHOT: checking for updates from mirror.snapshots
[INFO] snapshot tele.persistance:pers-api:1.2.0-alpha-1-SNAPSHOT: checking for updates from mirror.snapshots
[INFO] snapshot tele.persistance:pers-commons:1.2.0-alpha-1-SNAPSHOT: checking for updates from mirror.snapshots
Downloading: http://proxy/maven2-snapshots/repository/tele/persistance/pers-commons/1.2.0-alpha-1-SNAPSHOT/pers-commons-1.2.0-alpha-1-SNAPSHOT.jar
[WARNING] Unable to get resource 'tele.persistance:pers-commons:jar:1.2.0-alpha-1-SNAPSHOT' from repository mirror.snapshots (http://proxy/maven2-snapshots/repository)
[INFO] ------------------------------------------------------------------------
[ERROR] BUILD ERROR
[INFO] ------------------------------------------------------------------------
[INFO] Failed to resolve artifact.

Missing:
----------
1) tele.persistance:pers-commons:jar:1.2.0-alpha-1-SNAPSHOT

  Try downloading the file manually from the project website.

  Then, install it using the command: 
      mvn install:install-file -DgroupId=tele.persistance -DartifactId=pers-commons \
          -Dversion=1.2.0-alpha-1-SNAPSHOT -Dpackaging=jar -Dfile=/path/to/file

  Path to dependency: 
  	1) tele.persistance:pers-dao:jar:1.2.0-alpha-1-SNAPSHOT
  	2) tele.persistance:pers-commons:jar:1.2.0-alpha-1-SNAPSHOT

----------
1 required artifact is missing.

for artifact: 
  tele.persistance:pers-dao:jar:1.2.0-alpha-1-SNAPSHOT

from the specified remote repositories:
  central (http://repo1.maven.org/maven2),
  mirror.snapshots (http://proxy/maven2-snapshots/repository)
```

If I make an intermediate "install", everything works fine

---

**Affects:** 2.2

**Attachments:**
- [clean javadoc-plugin-test-case with classifier use.zip](https://issues.apache.org/jira/secure/attachment/12716919/clean+javadoc-plugin-test-case+with+classifier+use.zip) (_5.09 kB_)
- [javadoc-plugin-test-case.zip](https://issues.apache.org/jira/secure/attachment/12716911/javadoc-plugin-test-case.zip) (_5.89 kB_)
- [javadoc-plugin-test-case with classifier use.zip](https://issues.apache.org/jira/secure/attachment/12717060/javadoc-plugin-test-case+with+classifier+use.zip) (_157.84 kB_)
- [log.txt](https://issues.apache.org/jira/secure/attachment/12717027/log.txt) (_404.21 kB_)
- [mymobiliartest.zip](https://issues.apache.org/jira/secure/attachment/12716969/mymobiliartest.zip) (_47.58 kB_)
- [tiles-log.txt](https://issues.apache.org/jira/secure/attachment/12717028/tiles-log.txt) (_14.99 kB_)

**Issue Links:**
- [MJAVADOC-437](https://issues.apache.org/jira/browse/MJAVADOC-437) javadoc:aggregate fails on initial build
 (_**"duplicates"**_)
- [MJAVADOC-260](https://issues.apache.org/jira/browse/MJAVADOC-260) Javadoc report not generated for multi-module project if run from parent level - error resolving dependencies
 (_**"is duplicated by"**_)
- [MJAVADOC-264](https://issues.apache.org/jira/browse/MJAVADOC-264) mvn clean install site doesn't work
 (_**"is duplicated by"**_)
- [MJAVADOC-437](https://issues.apache.org/jira/browse/MJAVADOC-437) javadoc:aggregate fails on initial build

- [MJAVADOC-119](https://issues.apache.org/jira/browse/MJAVADOC-119) Aggregate does not work for multiple module project when doing install.

- [MNG-3685](https://issues.apache.org/jira/browse/MNG-3685) Dependency can't be resolved but has been found in the reactor.

- [MJAVADOC-276](https://issues.apache.org/jira/browse/MJAVADOC-276) Initial builds of a multi-module project fail

- [MJAVADOC-111](https://issues.apache.org/jira/browse/MJAVADOC-111) Can't release when I use javadoc aggregate/groups

- [MNG-3023](https://issues.apache.org/jira/browse/MNG-3023) Reactor projects should be included in dependency resolution

- [MNG-3283](https://issues.apache.org/jira/browse/MNG-3283) Plugins that require dependency resolution in early phases cause dependency resolution issue
 (_**"depends upon"**_)

51 votes, 44 watchers

**Comment by jira-importer** [comment]
**[Antonio Petrelli](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=apetrelli)** commented

The same problem appears during the use of the release plugin: the release fails and I have to install the checked-out code locally before retrying the release.

**Comment by jira-importer** [comment]
**[Arne Degenring](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=arned)** commented

I face the same problem. It seems to be a regression, as the problem does not occur with maven-javadoc-plugin 2.0.

**Comment by jira-importer** [comment]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** commented

Same thing with releases version of artifacts.

This is a major issue in a multi-module project when you want to release the project and the aggregated javadoc.

As the maven-release-plugin does "deploy site-deploy" when release:perform, this step always fails because of this bug.

**Comment by jira-importer** [comment]
**[Antonio Petrelli](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=apetrelli)** commented

Attached a small test case.
I noticed that the bug is triggered if a module depends on another, even if in correct order.
This bug is still present in the 2.3 version of the plugin.

**Comment by jira-importer** [comment]
**[Antonio Petrelli](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=apetrelli)** commented

Could someone add the "Testcase included" flag?
Thanks.

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

testcase included flag to yes.

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

Fixed in r617996.
Snapshot deployed but you need to have maven 2.0.8 to run it.

**Comment by jira-importer** [comment]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** commented

I tried with Maven 2.0.8 and maven-javadoc-plugin-2.4-20080203.152011-3

```
[INFO] Generate "JavaDocs" report.
[INFO] snapshot tele.persistance:pers-commons:jar:1.3.0-M1-SNAPSHOT: checking for updates from apache.snapshots
Downloading: http://people.apache.org/maven-snapshot-repository/tele/persistance/pers-commons/1.3.0-M1-SNAPSHOT/pers-commons-1.3.0-M1-SNAPSHOT.jar
[WARNING] The dependency: [tele.persistance:pers-commons:jar:1.3.0-M1-SNAPSHOT} can't be resolved but has been found in the reactor (probably snapshots).
This dependency has been excluded from the Javadoc classpath. You should rerun javadoc after executing mvn install.
[WARNING] IGNORED to add some artifacts in the classpath. See above.

...

[INFO] ------------------------------------------------------------------------
[ERROR] BUILD ERROR
[INFO] ------------------------------------------------------------------------
[INFO] Error during page generation

Embedded error: Error rendering Maven report: Missing:
----------
1) tele.persistance:pers-test:jar:tests:1.3.0-M1-SNAPSHOT

  Try downloading the file manually from the project website.

  Then, install it using the command: 
      mvn install:install-file -DgroupId=tele.persistance -DartifactId=pers-test -Dversion=1.3.0-M1-SNAPSHOT -Dclassifier=tests -Dpackaging=jar -Dfile=/path/to/file

...
```

I don't know if javadoc is correctly generated when I see all those WARNING messages.
Then build fails because of a dependency with "tests" classifier.

2 questions :

- when a jar file located in reactor is excluded from javadoc, are corresponding project's classes still included in javadoc ? (aggregate = true)

- are classifiers handled by reactor jar exclusion ?

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

Reopen since checkMissingArtifactsInReactor() doesnt take care of classifier.

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

Damien, did you call mvn site or mvn javadoc:javadoc?
Could you send us a log with -X?

**Comment by jira-importer** [comment]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** commented

> did you call mvn site or mvn javadoc:javadoc?

mvn -DreleaseProfile=true clean deploy site-deploy

So both site and javadoc:javadoc together.

Log file is coming...

**Comment by jira-importer** [comment]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** commented

Log file for mvn -X -DreleaseProfile=true clean deploy site-deploy in attachement.

Good luck

**Comment by jira-importer** [comment]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** commented

Vincent, do you need more information to fix the problem ?

**Comment by jira-importer** [comment]
**[Wendy Smoak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wsmoak)** commented

"mvn -DreleaseProfile=true clean deploy site-deploy" works for me with the latest maven-javadoc-plugin snapshot and the attached example.

(Well, after adding \<distributionManagement>   - for future examples, you can use something like file:///tmp/repo-MJAVADOC-116 as the deployment repo url, and likewise for the site deployment url.)

I'm not sure of the timing, but this _may_ have been fixed by along with MJAVADOC-137 by reverting the change from MJAVADOC-104 that added `@aggregator`.

**Comment by jira-importer** [comment]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** commented

Still doesn't work with classifiers

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

Damien, it seems that it is due to `@requiresDependencyResolution`, not the Javadoc plugin.
Could you confirm that the error comes from DefaultPluginManager ?

**Comment by jira-importer** [comment]
**[Wendy Smoak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wsmoak)** commented

As mentioned, I can't reproduce this with the attached example.

I'd like to stage the 2.4 release on Friday.

In order for this to be included, I need a sample project and steps to reproduce the problem.  Otherwise it will either be bumped to 2.5 or closed as Cannot Reproduce.  Thanks!

**Comment by jira-importer** [comment]
**[Antonio Petrelli](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=apetrelli)** commented

Wendy, can you wait one day? I tried some days ago with Tiles and it didn't work.

**Comment by jira-importer** [comment]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** commented

Here is a test case which reproduces problem with classifiers and aggregated javadoc.

It is based on previous patch with following updates :
- javadoc version 2.4-SNAPSHOT
- build module 1 test jar ("test-jar" classifier)
- module 2 depends on module 1 jar AND test-jar. They have to be declared together.

Try "mvn -DreleaseProfile=true clean deploy site-deploy", it will fail (maven 2.0.8)

Things to notice :
- with aggregate = false, it works
- in module 2, with exactly one of the 2 dependencies (jar OR test-jar), it works (??)

Is it enough to reproduce bug ?

**Comment by jira-importer** [comment]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** commented

Same test cas as before, but without any target folders.

**Comment by jira-importer** [comment]
**[Antonio Petrelli](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=apetrelli)** commented

Creating Javadocs with:
mvn javadoc:javadoc
does not work with Tiles:
http://svn.apache.org/repos/asf/tiles/framework/trunk/

Added tiles-log.txt as attachment.

**Comment by jira-importer** [comment]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** commented

Antonio, same problem as me with classifiers.

**Comment by jira-importer** [comment]
**[dlecan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dlecan)** commented

Wendy, this fix is waited for a long time by many people and is really a major issue.
Is it really not possible to work on this for 2.4 ?

**Comment by jira-importer** [comment]
**[Geoffrey Wiseman](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=diathesis)** commented

This is a real thorn in our sides.  Seems like we may have to disable this plugin.

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

MJAVADOC-197 could be a workaround for this issue.

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

MJAVADOC-197 is the only workaround that we could propose at this time.

**Comment by jira-importer** [comment]
**[Antonio Petrelli](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=apetrelli)** commented

Closed as fixed?
This bug is not fixed, there is only a workaround!
And anyway, we need to publish aggregated javadoc in Tiles when releasing it through the release plugin.
Our workaround is installing the tagged version (tagged using release:prepare) before running release:perform.

So please reopen this bug, or close it as "won't fix", "fixed" is misleading.

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

As requested

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

As requested on dev@ by Antonio

**Comment by jira-importer** [comment]
**[Shibu Gope](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shibug)** commented

It is still failing the build if javadoc is run in aggregate mode.  I set the \<failOnError>false\</failOnError> but it still fails.

**Comment by jira-importer** [comment]
**[Patrick M.J. Roth](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmjroth)** commented

I already tried

\<execution>
\<phase>verify\</phase>
\</execution>

but it didn't solve the issue.

As far as I have seen, a component (javadoc plugin?, site plugin? or more probably the maven framework itself) checks the dependencies very early in  the life cycle, independently of the phase at which the javadoc is executed.

I also tried to build with --fail-never. In that case modules are built, but the build itself is reported as failed. The site itself is not generated.

However the following works:
mvn install javadoc:aggregate

This could mean that the javadoc plugin is not involved.

**Comment by jira-importer** [comment]
**[Patrick M.J. Roth](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pmjroth)** commented

Sorry, you're right Emmanuel, 'verify' is before 'install'. But I also tried 'install', with no success.

I have debugged what happens in my case. I've seen that an exception is thrown when the jar can't be downloaded. In the catch clause it is then checked if the jar is part of the project. Since it is the case, the exception is ignored and a warning is written (something like "can't find dependency, but appears to be part of reactor, try to go further").

As you see, maven can recover from such an error. However, a file with extension .lastUpdated is written in place of the jar not found. If the same jar is searched for once more, then an new error is thrown and the build fails definitely.

I didn't analyze further, because it apears to me to be quite complex. Furthermore, I can circumvent the problem by running 'install' and 'site' in two separate commands.
